agents/craftassist/make_swarm.py

```diff
@@ -0,0 +1,367 @@
+                                         memory_receive_queue=self.memory_receive_queue,
+                                         memory_tag="swarm_worker_{}".format(agent.agent_idx))        
+        # controller
+        agent.disable_chat = True
+    
+    def check_task_info(self, task_name, task_data):
+        if task_name not in self.TASK_INFO.keys():
+            logging.info("task {} received without checking arguments")
+            return True
+        for key in self.TASK_INFO[task_name.lower()]:
+            if key not in task_data:
+                return False
+        return True
+
+    def preprocess_data(self, task_name, task_data):
+        if "task_data" in task_data:
+            return task_data["task_data"]
+        else:
+            return task_data
+
+    def send_task_updates(self, task_updates):
+        """send task updates to master by pushing to self.query_from_worker
+        """
+        if len(task_updates)>0:
+            name = 'task_updates'
+            self.query_from_worker.put((name, task_updates))
+
+    def handle_input_task(self, agent):
+        flag = True
+        while flag:
+            if self.input_tasks.empty():
+                flag = False
+            else:
+                task_class_name, task_data, task_memid = self.input_tasks.get_nowait()
+                if task_class_name not in self.TASK_MAP.keys():
+                    logging.info("task not understood by worker")
+                    continue
+                if task_memid is None or ((task_memid not in agent.task_stacks.keys()) and (task_memid not in agent.task_ghosts)):
+                    task_data = self.preprocess_data(task_class_name, task_data)
+                    if self.check_task_info(task_class_name, task_data):
+                        new_task = self.TASK_MAP[task_class_name](agent, task_data)
+                        if task_memid is None:
+                            task_memid = new_task.memid
+                        else:
+                            agent.task_ghosts.append(new_task.memid)
+                            # can send updates back to main agent to mark as finished
+                        agent.task_stacks[task_memid] = new_task
+                        agent.prio[task_memid] = -1
+                        agent.running[task_memid] = -1
+                        agent.pause[task_memid] = False
+                elif task_memid in agent.task_stacks.keys():
+                    self.send_task_updates([(task_memid, (agent.prio[task_memid], agent.running[task_memid], agent.task_stacks[task_memid].finished))])
+                elif task_memid in agent.task_ghosts:
+                    self.send_task_updates([(task_memid, (0, 0, True))])
+    
+    def handle_master_query(self, agent):
+        flag = True
+        while flag:
+            if self.query_from_master.empty():
+                flag = False
+            else:
+                query_name, query_data = self.query_from_master.get_nowait()
```

Comment:
not sure I understand the flow here - does this PR only handle "stop" and "resume" . How about other queries ?

Reply:
nope. Other queries are handled as well. It's just stop and resume are explicitly handled